typescript-test-samples/apigw-lambda-sqs-sns-datadog/README.md

```diff
@@ -0,0 +1,76 @@
+# Testing Serverless architectures with Datadog Synthetic Testing and Monitoring
+
+This example is about creating Synthetic Tests and Monitors, to ensure Serverless services are functioning properly in a real, AWS environment.
+
+Datadog makes it easy to add logging, metrics, distributed tracing, profiling, and more to Serverless architectures built on AWS.
+
+For this testing experience we'll use API Gateway, SNS, SQS, and Lambda:
+![](./images/architecture.png)
+
+
+## Setup
+
+1. Pull this repository
+2. Navigate to this example
+3. Run `npm install`
+4. Sign up for free [Datadog](https://datadoghq.com) trial
+5. Create an [API Key](https://app.datadoghq.com/account/login?next=%2Forganization-settings%2Fapi-keys)
+6. Deploy this example to AWS! Run `DD_API_KEY=<Your API Key> cdk deploy`
+
+## Testing
+1. Manually curl the endpoint to verify things work:
```

Comment:
Getting Internal server error here.

CloudWatch logs for the Lambda function show this:

```2023-05-08T18:11:23.610Z	undefined	ERROR	Uncaught Exception 	{
    "errorType": "Runtime.ImportModuleError",
    "errorMessage": "Error: Cannot find module 'publisher'\nRequire stack:\n- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js\n- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js",
    "stack": [
        "Runtime.ImportModuleError: Error: Cannot find module 'publisher'",
        "Require stack:",
        "- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js",
        "- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js",
        "    at ImportModuleError.ExtendedError [as constructor] (/opt/nodejs/node_modules/datadog-lambda-js/runtime/errors.js:113:28)",
        "    at new ImportModuleError (/opt/nodejs/node_modules/datadog-lambda-js/runtime/errors.js:123:42)",
        "    at /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:273:31",
        "    at step (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:43:23)",
        "    at Object.throw (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:24:53)",
        "    at rejected (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:16:65)"
    ]
}
```

Reply:
When I run the test in APIGW I get this error:

```
Mon May 08 18:21:22 UTC 2023 : Lambda execution failed with status 200 due to customer function error: Error: Cannot find module 'axios'
Require stack:
- /var/task/publisher.js
- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js
- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js. Lambda request id: 99657aac-2f6e-4c27-8852-7b36183703fa
```

Reply:
Ah looks like `npm install` and `npm run build` won't install the dependencies in the `resources` directory. Should I lift axios to be a top level dependency then? I thought that CDK's `assetFrom` method required all deps to be in the same directory. Maybe there's a canonical fix for this here?

This should work if you run `npm install` manually in the `resources` directory.

Reply:
Running `npm install` from the `resources` directory did the trick for me.

As far as a "canonical fix"....

This pattern uses CDK, we could potentially use a similar config:
https://github.com/aws-samples/serverless-test-samples/tree/main/typescript-test-samples/lambda-handler-dynamodb

IIRC its based off of `cdk init` and doesn't use a separate `resources` package - just the top-level one. Personally, I've used way more SAM than CDK, so I'll ask internally for another pair of eyes.

Reply:
Okay, I can also update the readme or even use `fetch` instead of axios. I'm also not extremely familiar with the CDK and mostly use the Serverless Framework, so it's a bit unclear in this context how to differentiate from runtime dependencies from build time dependencies - however if putting all of the application code at the root level works, we could do that.

Let me know, thanks!

Reply:
CDK best practices include [keeping infra and runtime code in the same package](https://docs.aws.amazon.com/cdk/v2/guide/best-practices.html#best-practices-constructs).

CDK init puts runtime code in a subdirectory, but uses shared root-level dependencies (single `package.json` in the root).

To avoid huge functions, the CDK `aws-cdk-lib/aws-lambda-nodejs/NodejsFunction` construct uses `esbuild` (or Docker) to tree-shake and bundle individual functions.

Suggestions here (will comment on the specific files as well for clarity):

1. Remove `/resources/package.json`
2. In `/lib/apigw-lambda-sqs-sns-datadog-stack.ts`, import `NodejsFunction` and use it instead of `lambda.Function` as seen [in this example](https://github.com/aws-samples/serverless-test-samples/blob/076d8652936124cabf1e1a98a9c9b47f215d3dff/typescript-test-samples/lambda-handler-dynamodb/lib/lambda-handler-dynamodb-stack.ts#L8)
    1. Set `entry` property instead of `code` property
    2. Adjust `handler` property to drop module name - e.g. `handler: "main"` instead of `handler: "publisher.main"`
5. Confirm `./package.json` includes axios (looks like it does at the moment)
6. Install a devDependency on `esbuild` to [avoid requiring Docker](https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-nodejs-readme.html#local-bundling)

These tweaks keep the bulk of the app structure as-is, which is aligned to CDK recommended structure, and also allow for the nice and simple README instructions you have already.